devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java

```diff
@@ -131,216 +90,65 @@ public Path resolve(AppArtifact appArtifact) throws AppModelResolverException {
-            boolean firstLevelExt,
-            Set<AppArtifactKey> visited) {
-        for (ResolvedDependency dep : resolvedDeps) {
-            final AppArtifactKey key = new AppArtifactKey(dep.getModuleGroup(), dep.getModuleName());
-            if (!visited.add(key)) {
-                continue;
-            }
-            final AppDependency appDep = versionMap.get(key);
-            if (appDep == null) {
-                // not a jar
-                continue;
-            }
-
-            Dependency extDep = null;
+        for (io.quarkus.gradle.workspace.model.Dependency extensionDependency : model.getExtensionDependencies()) {
+            AppDependency appDep = toAppDependency(extensionDependency);
             for (Path artifactPath : appDep.getArtifact().getPaths()) {
                 if (!Files.exists(artifactPath)) {
                     continue;
                 }
                 if (Files.isDirectory(artifactPath)) {
-                    extDep = processQuarkusDir(appDep.getArtifact(), artifactPath.resolve(BootstrapConstants.META_INF),
+                    processQuarkusDir(appDep.getArtifact(), artifactPath.resolve(BootstrapConstants.META_INF),
                             appBuilder);
                 } else {
                     try (FileSystem artifactFs = FileSystems.newFileSystem(artifactPath, null)) {
-                        extDep = processQuarkusDir(appDep.getArtifact(), artifactFs.getPath(BootstrapConstants.META_INF),
+                        processQuarkusDir(appDep.getArtifact(), artifactFs.getPath(BootstrapConstants.META_INF),
                                 appBuilder);
                     } catch (IOException e) {
                         throw new GradleException("Failed to process " + artifactPath, e);
                     }
                 }
-                if (extDep != null) {
-                    break;
-                }
             }
-
-            boolean addChildExtensions = firstLevelExt;
-            if (extDep != null && firstLevelExt) {
-                firstLevelExtensions.add(extDep);
-                addChildExtensions = false;
+            if (!userDeps.contains(appDep)) {
+                deploymentDeps.add(appDep);
             }
-            final Set<ResolvedDependency> resolvedChildren = dep.getChildren();
-            if (!resolvedChildren.isEmpty()) {
-                collectExtensionDeps(resolvedChildren, versionMap, appBuilder, firstLevelExtensions, addChildExtensions,
-                        visited);
+        }
+
+        final List<AppDependency> fullDeploymentDeps = new ArrayList<>(userDeps);
+        fullDeploymentDeps.addAll(deploymentDeps);
+
+        if (!appArtifact.isResolved()) {
+            PathsCollection.Builder paths = PathsCollection.builder();
+            for (SourceSet sourceSet : model.getSourceSets()) {
+                sourceSet.getSourceDirectories().stream().filter(File::exists).map(File::toPath).forEach(paths::add);
```

Comment:
Are these java sources?

Reply:
these are more `BuildOutput` such as `build/classes/java/main`

Reply:
well actually it's not, but I will update so it points to classes and not java files